[IndexProvider/Table] Migrate from web - take 2

AndrewMusgrave opened the pull request:
### WHY are these changes introduced?

Feature requests: [#390, #3637]


As teams reach AODA they'll need to replace their resource list with headers, with a semantic table. This pull request is bringing the table from orders into Polaris in an unstable state so others can gain from its accessibility improvements and contribute to the stability of its API/name

### Major hanges other than C+P

* Add hook to streamline component usage
<img width="507" alt="Screen Shot 2020-11-18 at 4 46 02 PM" src="https://user-images.githubusercontent.com/24610840/99592068-ad6b9a00-29bd-11eb-8760-b8b7c17592e3.png">

* Add negative margin while in select mode with bulk actions to remove extra white space at the bottom of the list
* Add tests
* Change loading bar colors to meet contrast requirements

### Notes

Accessibility tests are failing from the filter's component - going to address them in another pull request

### WHAT is this pull request doing?

<!--
  Summary of the changes committed.

  Before / after screenshots are appreciated for UI changes. Make sure to include alt text that describes the screenshot.

  If you include an animated gif showing your change, wrapping it in a details tag is recommended. Gifs usually autoplay, which can cause accessibility issues for people reviewing your PR:

    <details>
      <summary>Summary of your gif(s)</summary>
      <img src="..." alt="Description of what the gif shows">
    </details>
-->

## <!-- ℹ️ Delete the following for small / trivial changes -->

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

<!--
  Give as much information as needed to experiment with the component
  in the playground.
-->

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}

```

</details>

### Screenies

#### Simple index table
<img width="717" alt="Screen Shot 2020-11-18 at 4 53 16 PM" src="https://user-images.githubusercontent.com/24610840/99592732-94171d80-29be-11eb-93bd-c41285af54ee.png">

#### Simple small screen index table
![Screen Shot 2020-11-18 at 4 48 58 PM](https://user-images.githubusercontent.com/24610840/99592607-6f22aa80-29be-11eb-8b8f-430776a7334c.png)

#### Index table with empty state
![Screen Shot 2020-11-18 at 4 49 05 PM](https://user-images.githubusercontent.com/24610840/99592614-70ec6e00-29be-11eb-93ef-7a693826421d.png)

#### Index table with bulk actions
![Screen Shot 2020-11-18 at 4 49 15 PM](https://user-images.githubusercontent.com/24610840/99592624-72b63180-29be-11eb-9135-30bd54274b18.png)

#### Index table with bulk actions across pages
![Screen Shot 2020-11-18 at 4 49 24 PM](https://user-images.githubusercontent.com/24610840/99592636-76e24f00-29be-11eb-87b0-c3e8e7d2d699.png)

#### Index table with loading state
![Screen Shot 2020-11-18 at 4 49 28 PM](https://user-images.githubusercontent.com/24610840/99592647-7944a900-29be-11eb-90e4-ed9085359cb5.png)

#### IndexTable with filtering
![Screen Shot 2020-11-18 at 4 49 49 PM](https://user-images.githubusercontent.com/24610840/99592653-7b0e6c80-29be-11eb-86bb-a683c7d2cabd.png)

#### IndexTable with all of its elements
![Screen Shot 2020-11-18 at 4 49 55 PM](https://user-images.githubusercontent.com/24610840/99592660-7cd83000-29be-11eb-81d9-5ff0585afb54.png)

#### Small screen IndexTablr with all of its elements
![Screen Shot 2020-11-18 at 4 50 05 PM](https://user-images.githubusercontent.com/24610840/99592667-7e095d00-29be-11eb-84fc-26c616977796.png)

AndrewMusgrave commented on src/components/IndexTable/IndexTable.scss:
```diff
@@ -72,7 +72,7 @@ $loading-panel-height: rem(53px);
 .LoadingPanelText {
   margin-left: rem(12px);
-  color: var(--p-text-highlight, color('teal', 'dark'));
+  color: var(--p-text, color('teal', 'darker'));
```
Had to adjust colors to meet contrast requirements. We'll want to take a look at making `--p-text-highlight` accessible.

AndrewMusgrave commented:
Other accessibility tests are fixed in : https://github.com/Shopify/polaris-react/pull/3651

AndrewMusgrave commented on .storybook/polaris-readme-loader.js:
```diff
@@ -138,8 +138,8 @@ import {
   Image,
-  IndexProvider,
-  IndexTable,
+  UnstableIndexProvider as IndexProvider,
```
Commit adding unstable prefix

AndrewMusgrave replied:
The instability is regarding the API since we've never revisited it. I was hoping to take a look at it this week but didn't have the chance. I'm going on vacation but ill take a look when I'm back and see if we can resolve the API

AndrewMusgrave commented on .storybook/polaris-readme-loader.js:
```diff
@@ -194,7 +194,7 @@ import {
   VisuallyHidden,
   VideoThumbnail,
-  useIndexResourceState,
+  useIndexResourceStateUnstable as useIndexResourceState,
```
Used suffix for the hook - since hooks always start with `use` as per convention

dfmcphee commented on src/components/IndexProvider/IndexProvider.tsx:
```diff
@@ -0,0 +1,233 @@
+    <IndexContext.Provider value={contextValue}>
+      <IndexRowContext.Provider value={rowContextValue}>
+        <IndexSelectionChangeContext.Provider
+          value={handleSelectionChange as any}
```
Why is the `any` needed here?

dfmcphee commented on src/components/IndexTable/IndexTable.tsx:
```diff
@@ -0,0 +1,619 @@
+    tableHeadingRects.current = measuredTableHeadingRects;
+  }, []);
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
If you convert this to `useMemo` instead, you shouldn't need the eslint disable like we do [elsewhere](https://github.com/Shopify/polaris-react/blob/master/src/components/ActionMenu/components/Actions/Actions.tsx#L120).

dfmcphee commented on src/components/IndexTable/README.md:
```diff
@@ -0,0 +1,1244 @@
+
+A index table with simple items and no bulk actions, sorting, or filtering.
+
+TODO
```
What is left TODO here?

dfmcphee commented on src/components/IndexTable/README.md:
```diff
@@ -0,0 +1,1244 @@
+    selectedResources,
+    allResourcesSelected,
+    handleSelectionChange,
+  } = useIndexResourceState(customers);
```
This is so much better 🥇

dfmcphee commented on src/components/IndexTable/README.md:
```diff
@@ -0,0 +1,1244 @@
+
+### Simple small screen index table
+
+A small screen index table with simple items and no bulk actions, sorting, or filtering.
```
I wonder if rather than having a separate example like this that looks strange when viewed on larger screens, we instead have one example that actually does the responsive switch between the views. I think that would be more helpful to consumers who are trying to wire this up.

AndrewMusgrave replied:
@umar-k Is adding responsive examples [here](https://github.com/Shopify/polaris-react/pull/3951) 🎉

Alexzanderk commented:
@alex-page @dfmcphee @translation-platform Hi guys! How is it going with this PR?

AndrewMusgrave commented on tests/setup.ts:
```diff
@@ -7,6 +7,13 @@ configure({adapter: new Adapter()});
 // Mocks for scrolling
 window.scroll = () => {};
 
+window.open = (
```
Adding a mock for `window.open`

AndrewMusgrave commented on src/components/IndexTable/tests/IndexTable.test.tsx:
```diff
@@ -23,18 +23,33 @@ const mockTableItems = [
   return <Component {...item} key={item.id} />;
 };
 
+const mockRenderCondensedRow = (item: any) => {
```
Adding a second renderer to remove the console errors

AndrewMusgrave commented on src/components/IndexTable/components/Row/tests/Row.test.tsx:
```diff
@@ -41,13 +41,9 @@ describe('<Row />', () => {
 
   it('renders a RowHoveredContext provider', () => {
     const row = mountWithTable(
-      <table>
```
We're already using `mountWithTable`

AndrewMusgrave commented:
@dfmcphee I took a look at Checkbox, Row & Cell.

Row didn't need any changes. 

For `Cell` & `Checkbox`, I removed the `last` & `sub` props since they're not used in production. I changed `noPadding` to `flush` to match Modal's API. `first` has also been removed. The styles are being applied through a selector rather than a second class. `Row` now renders a checkbox by default, allowing us to remove IndexTable.Checkbox from the API assisting in small screen view creation.

I briefly looked at IndexTable & IndexProvider. I removed `onMoreActionPopoverToggle` for now till we're more confident about it. On Friday I'm hoping to finish reworking these. I want to take a look changing the `sort` prop on IndexTable and see if we have a better way of handling it. We can also remove `resolveItemId` from 'IndexProvider` & I'm thinking we'll be able to combine the components since we don't get a lot of value from having them separate.

What do you think?

dfmcphee commented:
That all sounds great to me. I think combining makes sense, at least for consumers right now. It still would be nice to be able to have some of the index logic abstracted from the view for other types in the future (index grid for example). Could we keep the provider but have it automatically included in the table?

lsit commented:
Hi! Love this PR! We (Merchandising) have converted Products and Collections to use the `IndexTable` as well and we've collected some improvements [here](https://github.com/Shopify/web/issues/36701). If you need help testing or reviewing, I'm sure @xie or I could help out. Thanks for doing this! 🎉

GhostApps commented:
Is this going to be in v6?

AndrewMusgrave commented:
Hi @GhostApps, this will not be bundled with v6. V6 primarily revolves around design language and accessibility standards (AODA). This change would possibly delay the v6 rollout. Ideally, we'll have this ready shortly after and ship it out in a minor version!

Alexzanderk commented:
@AndrewMusgrave hi, do you have any deadlines for adding indexTable?

AndrewMusgrave commented on src/components/IndexTable/IndexTable.tsx:
```diff
@@ -38,7 +39,7 @@ export interface IndexTableHeading {
 
 export interface IndexTableBaseProps {
-  headings: IndexTableHeading[];
+  headings: NonEmptyArray<IndexTableHeading>;
```
Since we access headings via index then dot notation we'll want to type the array to have headings

AndrewMusgrave commented on src/components/IndexTable/IndexTable.tsx:
```diff
@@ -80,10 +80,6 @@ function IndexTableBase({
-  const {
-    value: hasMoreRightColumns,
-    toggle: toggleHasMoreRightColumns,
-  } = useToggle(false);
```
Removing this since it's not being used

AndrewMusgrave commented:
I've just finished adding test coverage for all of our reachable code (refs are unreachable). We'll want to comb through the components later to improve actual test coverage of the functionality but its miles ahead of what it was previously.

API has been slightly restricted. Props that were not being used have been removed. IndexTable is an internal component now while we decide how we want to handle it.

@lsit , @xie , @lhoffbeck, @CarlGranstrom  have been making a lot of great contributions to the table in web. We'll want to work on migrating the contributions back into the Polaris IndexTable to ease the migration progress and deprecation of the table in web. I think there's still some work to be done on the component (see @lsit list) but now that's it's in Polaris it'll be easier to get contributions 🎉 

@dfmcphee @alex-page @umar-k This is ready for review again!

alex-page reviewed (approved):
I am happy to get this into Polaris and iterate as needed. Would love to get the recent changes in web applied here as well so we can remove those components.

dfmcphee reviewed (approved):
This looks great. So excited to see this make its way into Polaris. 🎉

umar-k commented on src/components/IndexTable/IndexTable.tsx:
```diff
@@ -0,0 +1,630 @@
+    tableElement.current && tableInitialized
+      ? {
+          height: '1px',
+          width: tableElement.current.offsetWidth - SCROLL_BAR_PADDING,
```
Is `tableElement.current.offsetWidth` supposed to change with every render when the browser is being resized?
It seems to stay the same and might be whats causing the scrollbar issue.

AndrewMusgrave replied:
@lhoffbeck Sounds like this might be what you noticed this morning

lhoffbeck replied:
Yeah that looks exactly like what I was seeing!

AndrewMusgrave replied:
@umar-k Should be fixed now - [here's the commit](https://github.com/Shopify/polaris-react/pull/3646/commits/b529ce37b42a949d74b306e42d9163b36fd489b5)

umar-k reviewed (commented):
Component looks really good 👍 

I noticed one issue with the horizontal scrollbar during 🎩 
It looks like the `scrollBarContentStyles` doesn't get an updated width when the window is resized. It causes a scrollbar even when it's not needed.


https://user-images.githubusercontent.com/72987032/107058941-3adfae80-67a3-11eb-870e-eba55b2f31ed.mov

AndrewMusgrave commented:
We still need 3 required languages to 🚢 however the estimate is that they'll come in over the weekend.

umar-k reviewed (approved):
🎩 'd and looks good to me! 🎉

GhostApps commented:
@AndrewMusgrave any chance you know when this will be released? I tried to play with it by pulling latest but couldn't get the playground or building for a consuming project to work so am having to wait for the next official Polaris release

AndrewMusgrave commented:
@GhostApps The Polaris team will have more context, but they usually draft a release ~2 weeks. The last one was 11 days ago so it shouldn't be to long.

What error were you getting in the playground? Maybe we could sort that out.

AndrewMusgrave commented:
If you wanted to create an issue and ping me there I would have happy to answer (so everyone in this pull request doesn't get pinged 😄 )
